File: recoverbs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 //TC-O(N),SC-O(h)
 class Solution {
    TreeNode first;
    TreeNode second;
    TreeNode prev;
    public void recoverTree(TreeNode root) {
        if(root==null) return;
        // find the nodes of breaches by validating the BST and swap the nodes, perform inorder traversal 
        inorder(root); 
        int temp=first.val;
        first.val =second.val;
        second.val = temp;
        System.out.println(first.val);System.out.println(second.val);
        
    }
    public void inorder(TreeNode root){
        //basecase
        if(root==null) return;
        inorder(root.left);
        if(prev!=null && prev.val > root.val){
            if(first==null){
                first = prev;
                second = root;
            }
            else{
               second=root;
            }
        }
            prev=root;
            inorder(root.right);
        }
    }

File: RecoverBS.py
#TC: O(n)
#SC: O(h)

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Wrapper:

    def __init__(self):
        self.prev = TreeNode(float('-inf'))
        self.first = None
        self.second = None

class RecoverBSTIteration:
    
    def __swap(self, node1: TreeNode, node2: TreeNode) -> None:

        tempVal = node1.val
        node1.val = node2.val
        node2.val = tempVal
    
    def recoverTree(self, root: TreeNode) -> None:
        
        if (root == None):
            return
        
        wrapObj = Wrapper()
        
        stack = []
        currNode = root
        
        while (currNode != None):
            stack.append(currNode)
            currNode = currNode.left
         
        while (len(stack) > 0):
            
            nextMinNode = stack.pop()
            
            if (wrapObj.first == None and wrapObj.prev.val > nextMinNode.val):
                wrapObj.first = wrapObj.prev

            if (wrapObj.first != None and wrapObj.prev.val > nextMinNode.val):
                wrapObj.second = nextMinNode
            
            wrapObj.prev = nextMinNode
            
            if (nextMinNode.right != None):
                currNode = nextMinNode.right
                while (currNode != None):
                    stack.append(currNode)
                    currNode = currNode.left
        
        self.__swap(wrapObj.first, wrapObj.second)


class RecoverBSTRecursion:
    
    def __swap(self, node1: TreeNode, node2: TreeNode) -> None:
        tempVal = node1.val
        node1.val = node2.val
        node2.val = tempVal
        
    def __helper(self, root: TreeNode, wrapObj: Wrapper) -> None:
        
        #   base case
        if (root == None):
            return
        
        self.__helper(root.left, wrapObj)
        
        if (wrapObj.first == None and wrapObj.prev.val > root.val):
            wrapObj.first = wrapObj.prev

        if (wrapObj.first != None and wrapObj.prev.val > root.val):
            wrapObj.second = root

        wrapObj.prev = root
        
        self.__helper(root.right, wrapObj)
    
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if (root == None):
            return
        
        wrapObj = Wrapper()
        
        self.__helper(root, wrapObj)
        
        self.__swap(wrapObj.first, wrapObj.second)